### ISSUE TITLE
MVC argument resolver for SecurityContext

### ISSUE THREAD

**Issue by chrylis** [spring-projects/spring-security#3720]
Getting a reference to the `SecurityContext` is needed for safely updating information about the logged-in user, but there is no MVC argument resolver for it. I have a trivial implementation that I'm using in my own code and would be happy to prepare a PR for it if it's considered worth including in the mainline.

**Comment by rwinch** [comment]
@chrylis Thanks for the offer. This does sound somewhat appealing. However, I'm concerned about the usage of SecurityContext and setting a new `Authentication` directly as this can cause race conditions. From [the reference](http://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#tech-intro-sec-context-persistence)

> In an application which receives concurrent requests in a single session, the same SecurityContext instance will be shared between threads. Even though a ThreadLocal is being used, it is the same instance that is retrieved from the HttpSession for each thread. This has implications if you wish to temporarily change the context under which a thread is running. If you just use SecurityContextHolder.getContext(), and call setAuthentication(anAuthentication) on the returned context object, then the Authentication object will change in all concurrent threads which share the same SecurityContext instance. You can customize the behaviour of SecurityContextPersistenceFilter to create a completely new SecurityContext for each request, preventing changes in one thread from affecting another. Alternatively you can create a new instance just at the point where you temporarily change the context. The method SecurityContextHolder.createEmptyContext() always returns a new context instance.

I wonder if we can accomplish something similar without the problems I outlined above.

**Comment by chrylis** [comment]
In my case, at least, I'm explicitly *trying* to set the new `Authentication` globally (ideally, including across all sessions, as in spring-session#398. Is the concern that some users would expect the injected context to be request-scoped and thus might munge the rest of the session accidentally?

**Comment by rwinch** [comment]
@chrylis I'm not sure I understand. Can you please clarify?

**Comment by chrylis** [comment]
In my case, I have changed a user's permissions (cached in the session) and want to replace the `Authentication` object for all of that user's future requests to reflect the permission change. Altering the security context (which AIUI in Spring MVC writes through to the session) is *supposed* to modify all the concurrent requests.

The race condition you describe, if I'm interpreting correctly, applies to multiple incoming requests where the controller or other participant in the request handling is implementing a delegation or run-as, intending to restore the original `Authentication` before terminating but accidentally messing with the session state for simultaneous requests. I believe your concern is that making the `SecurityContext` trivial to inject will lead people to publish the temporary change globally rather than creating a new temporary context locally.

**Comment by rwinch** [comment]
> I believe your concern is that making the SecurityContext trivial to inject will lead people to publish the temporary change globally rather than creating a new temporary context locally.

Correct. It seems it would lead to improper usage in real world scenarios.

**Comment by chrylis** [comment]
I'm sure that the temporary case is used frequently, but I've actually used the permanent case multiple times but never the temporary one.

In the run-as case, I believe that the user needs to operate through the context holder anyway to set the new empty-and-then-filled context as the current context and then restore it, meaning that there is little advantage to having an annotation parameter that specifies injecting an empty context, since the resolver could theoretically magically create and then set the empty context but couldn't restore it afterward.

This appears to come down to a judgment call on the OOTB convenience versus likelihood of subtle errors. I personally think that the exact same error is nearly as likely with `SecurityContextHolder.getContext().setAuthentication(a)`, but you're the one with the wide perspective.

**Comment by rwinch** [comment]
@chrylis I guess what I'm trying to get at is that we could support something like this, but in a less error prone way.

Perhaps we can allow a user to return an `Authentication` object and mark the method with `@AuthenticatedWith`. Spring Security would have a `ReturnValueHandler` that would then populate the `SecurityContextHolder` properly. It could also send the user to the saved request, etc. 

There may be times where the user doesn't want to go through a login flow. These instances we might have some value object that contains the `Authentication`. An alternative is to allow the `Authentication` to be populated in the model and allow `SpringSecurity` to process a specific model attribute.

These aren't ideas that are set in stone (i.e. I think that we can improve on names, how it works, etc). However, I think these are a step in the right direction.

**Comment by chrylis** [comment]
> Perhaps we can allow a user to return an Authentication object and mark the method with @AuthenticatedWith.

I'm looking at MVC request handlers here; the return value is already "taken" by the HTTP response in one form or another. In the instant case, the user is redirected back to the "account view" page, and if this part gets exposed as a JSON API, the user will get some form of plan-status object back. The authentication refresh necessarily happens as a side effect.
